README.md
# Analyse DLS with CONTIN (for Python)

[![Testing](https://github.com/BAMresearch/analyse_dls_with_contin/actions/workflows/tests.yml/badge.svg?branch=main)](https://github.com/BAMresearch/analyse_dls_with_contin/actions/workflows/tests.yml)
[![Documentation](https://github.com/BAMresearch/analyse_dls_with_contin/actions/workflows/doc.yml/badge.svg?branch=main)](https://github.com/BAMresearch/analyse_dls_with_contin/actions/workflows/doc.yml)

This repository contains Python code and a Jupyter Notebook
running the original [CONTIN program by S. Provencher](http://dx.doi.org/10.1016/0010-4655(82\)90174-6)
on every DLS measurement (dynamic light scattering, aka. photon correlation spectroscopy, PCS)
read from `*.ASC` files at the specified angle found in the given subfolders.
The expected input file format is `ALV-7004 CGS-8F Data` which is found at the first line of each file.

### Online Documentation

https://bamresearch.github.io/analyse_dls_with_contin

#### Documentation as PDF

https://bamresearch.github.io/analyse_dls_with_contin/analyse-dls-with-contin.pdf


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.1 (2023-02-23)
### Fix
* **Notebook:** Data path reset to included testdata, updated results ([`ab0f522`](https://github.com/BAMresearch/analyse_dls_with_contin/commit/ab0f522b078ce9d5e9bab8c30ca697316a213fbc))
* **dlshelpers:** Typo ([`81afc7a`](https://github.com/BAMresearch/analyse_dls_with_contin/commit/81afc7a35937c6d2457196a003ebf5ddfccdf350))


contin.py
# -*- coding: utf-8 -*-
# utils.py

import os, shutil, subprocess
import io
import time
import urllib
import requests
import itertools
from pathlib import Path
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np
import sys
cwd = Path().resolve()
if cwd not in sys.path:
    sys.path.insert(0, str(cwd))
from jupyter_analysis_tools.utils import isWindows, isMac, isList, pushd, grouper, updatedDict
from jupyter_analysis_tools.analysis import getModZScore
from dlshelpers import getDLSgammaSi, readDLSData

InputFn = "contin_in.txt"
OutputFn = "contin_out.txt"

def getContinForWindows(targetPath):
    binaryName = "contin-windows.exe"
    baseurl="http://www.s-provencher.com"
    html_page = urllib.request.urlopen(baseurl+"/contin.shtml").read()
    soup = BeautifulSoup(html_page)
    binurl = [link.get('href') for link in soup.findAll('a') if link.text.strip() == binaryName]
    binurl = '/'.join((baseurl,binurl[0]))
    binary = requests.get(binurl, allow_redirects=True)
    open(targetPath, 'wb').write(binary.content)
    return targetPath

def getContinPath():
    if isMac():
        # get local path to the CONTIN executable
        return Path.home() / "code" / "cntb2" / "bin" / "contin OSX"
    elif isWindows():
        continCmd = Path(os.getenv('APPDATA')) / "contin" / "contin.exe"
        #print(continCmd)
        if continCmd.is_file():
            return continCmd
        continCmd.parent.mkdir(parents=True, exist_ok=True)
        getContinForWindows(continCmd)
        if continCmd.is_file():
            print(f"Installed CONTIN at '{continCmd}'.")
        else:
            print("Failed to retrieve CONTIN executable!")
    raise NotImplementedError("Don't know how to retrieve the CONTIN executable!")

def genContinInput(filedata, **continConfig):
    """Expects a dictionary of file data created by readDLSData()."""
    IWT = 5 if continConfig['weighResiduals'] else 1
    # transform data? Trd=0: no transform
    Trd = -1 # Trd=1: initial g(2), input sqrt[g(2)-1]; Trd=-1: initial g(2)-1, input sqrt[g(2)-1]
    # select the measurement angle, make sure it's in the file
    angle = continConfig['angle']
    assert angle in filedata['angles'], \
        f"Given angle ({angle}) not found in file '{filedata['filename']}': {filedata['angles']}"
    # get environment values for storage in contin file
    temp    = filedata['Temperature [K]']
    visc    = filedata['Viscosity [cp]']
    refrac  = filedata['Refractive Index']
    wavelen = filedata['Wavelength [nm]']
    gamma = getDLSgammaSi(angle, refrac, wavelen*1e-9, temp, visc*1e-3)
    fitmin  = min(continConfig['fitRangeM'])/gamma
    fitmax  = max(continConfig['fitRangeM'])/gamma
    Im, dIm = 1, 0
    # get measured correlation data and tau
    dlsData = filedata["correlation"].reset_index()
    dlsData.tau *= 1e-3 # convert to seconds
    # restrict data to given range
    tmin, tmax = min(continConfig['ptRangeSec']), max(continConfig['ptRangeSec'])
    tmask = np.logical_and(tmin <= dlsData.tau, dlsData.tau <= tmax)
    tauCropped = dlsData.tau[tmask]
    corCropped = dlsData[angle][tmask]
    a2s_kwargs = dict(floatmode='fixed', sign=' ', max_line_width=80,
                      formatter={'float_kind': '{0: .5E}'.format})
    tauStr  = np.array2string(tauCropped.values, **a2s_kwargs)[1:-1]
    corStr  = np.array2string(corCropped.values, **a2s_kwargs)[1:-1]
    npts = len(tauCropped)
    # store the score for this data if it was determined
    scoreRecord = f"\n RUSER    11    {filedata['score'][angle]:.3E}" if 'score' in filedata else ""
    storedFn = filedata['filename'].name
    if not filedata['filename'].is_file():
        storedFn = filedata['filename'].parent.suffix + '/' + storedFn
    # generate CONTIN input file
    content = f"""{storedFn}
 IFORMY    0    .00
 (6E13.7)
 IFORMT    0    .00
 (6E13.7)
 IFORMW    0    .00
 (6E13.7)
 NINTT     0   -1.00
 DOUSNQ    0    1.00
 IQUAD          1.00
 PRWT      0    1.00
 PRY       0    1.00
 IPLRES    1    3.00
 IPLRES    2    3.00
 IPRINT    1    0.00
 IPRINT    2    2.00
 IPLFIT    1    0.00
 IPLFIT    2    0.00
 LINEPG    0   50.
 NONNEG    1    1.00
 IWT       0    {IWT:.2f}
 NQPROG    1    5.00
 NQPROG    2    5.00
 GMNMX     1    {fitmin:.3E}
 GMNMX     2    {fitmax:.3E}
 RUSER    10    {Trd:.2f}
 NG        0    {{gridpts:.2f}}
 NLINF     0    {{baselineCoeffs:.2f}}
 IUSER    10    4.00{scoreRecord}
 RUSER    21    1.0
 RUSER    22   -1.0
 RUSER    23    0.0
 RUSER    18         {temp:.5f}
 RUSER    17         {angle:.5f}
 RUSER    19         {visc:.5f}
 RUSER    15         {refrac:.5f}
 RUSER    16         {wavelen:.5f}
 RUSER    25         {Im}
 RUSER    26         {dIm}
 END       0    0.00
 NY{npts: >9d}
 {tauStr}
 {corStr}
""".format(**continConfig
    )
    return content.encode('ascii')

def getContinOutputDirname(angle):
    return f"contin_{angle:03.0f}"

def workerInit(_queue):
    """Initializes a queue for log messages in each worker process during multiprocessing.
    Queue object is global within each process only, not in the parent."""
    global queue
    queue = _queue

def runContin(filedata, continConfig, useQueue=True):
    """Starts a single CONTIN process for the given DLS DataSet
    (which should contain a single angle only)."""
    continCmd = getContinPath()
    assert continCmd.is_file(), "CONTIN executable not found!"
    logPrefix = f"{filedata['filename'].name}@{continConfig['angle']}°: "
    workDir = filedata['filename'].parent
    if workDir.is_file():
        logPrefix = workDir.stem + '/' + logPrefix
        workDir = workDir.parent / workDir.stem
    logFunc = queue.put if useQueue else print
    def log(text):
        #print("="+logPrefix+text)
        logFunc(" "+logPrefix+text)
    try:
        continInData = genContinInput(filedata, **continConfig)
    except AssertionError:
        log(f"Scattering angle {continConfig['angle']} not found! "
            f"Skipping…")
        return
    #ts = datetime.datetime.now().strftime("%Y%m%d-%H%M%S") # timestamp
    tmpDir = workDir / (getContinOutputDirname(continConfig['angle'])+' '+filedata['filename'].stem)
    if tmpDir.is_dir(): # deleting old results
        if not continConfig.get("recalc", True):
            return tmpDir
        shutil.rmtree(tmpDir)
    os.mkdir(tmpDir)
    continInDataPath  = tmpDir / InputFn
    continOutDataPath = tmpDir / OutputFn
    # Store input data
    with open(continInDataPath, 'wb') as fd:
        fd.write(continInData)
    with pushd(tmpDir):
        proc = subprocess.run([str(continCmd)], input=continInData,
                              stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        if len(proc.stderr):
            log(proc.stderr.decode().strip())
    # Store output data
    with open(continOutDataPath, 'wb') as fd:
        fd.write(proc.stdout)
    return tmpDir

def readData(fnLst, configLst):
    angles = [cfg['angle'] for cfg in configLst]
    dataLst = readDLSData(fnLst)
    # calc modified Z-Score based on median absolute deviation
    # for each count rate at the same angle
    for angle in set(angles):
        try:
            idx, cr = zip(*((i, filedata['countrate'][angle].values)
                             for i, filedata in enumerate(dataLst)
                                 if angle in filedata['countrate']))
        except (ValueError, KeyError): # filedata without 'countrate', most probably
            continue
        #print(angle, idx, getModZScore(np.stack(cr)))
        for i, score in zip(idx, getModZScore(np.stack(cr))):
            if 'score' not in dataLst[i]:
                dataLst[i]['score'] = dict()
            dataLst[i]['score'][angle] = score
    return dataLst

def runContinOverFiles(fnLst, configLst, nthreads=None, outputCallback=None):
    """*fnLst*: List of file paths to .ASC files
       *configLst*: List of parameters, one dict for each file, such as
            {'recalc': True, 'gridpts': 200, 'transformData': True,
             'ptRangeSec': (3e-07, 1.0), 'fitRangeM': (7e-10, 3.9e-07),
              'baselineCoeffs': 0, 'weighResiduals': True}
       *nthreads*: number of parallel CONTIN processes to launch,
           1: Sequential processing, one file after another
           None: number of processes equals the number of computing cores
       *outputCallback*: A function with one argument to called repeatedly (0.5s)
           with new output messages combined from all CONTIN processes.
    """
    start = time.time()
    if not isList(configLst):
        configLst = (configLst,)
    dataLst = readData(fnLst, configLst)
    # get all combinations of CONTIN parameters and data files
    dataNConfig = [(data, cfg) for data in dataLst for cfg in configLst]
    if nthreads == 1:
        resultDirs = [runContin(data, cfg, False) for data, cfg in dataNConfig]
    else: # Using multiple CPU cores if available
        import multiprocessing
        if not nthreads:
            nthreads = multiprocessing.cpu_count()
        from multiprocessing import Queue as MPQueue
        # use a queue to collect stdout messages from subprocesses
        logQueue = MPQueue()
        pool = multiprocessing.Pool(processes=nthreads, initializer=workerInit, initargs=(logQueue,))
        resultDirs = pool.starmap_async(runContin, dataNConfig)
        pool.close()
        def resultReady(asyncResult): # checks if the overall result is ready
            try:
                asyncResult.successful()
            except ValueError:
                return False
            return True
        outputBuffer = [] # buffer to store output messages from queue in,
                          # for sorting, for deterministic testing
        while not resultReady(resultDirs):
            time.sleep(.5) # update interval of output
            while not logQueue.empty():
                newOutput = logQueue.get_nowait()
                if not outputCallback:
                    print(newOutput) # the traditional way
                else:
                    outputBuffer.extend(newOutput.splitlines())
            if callable(outputCallback):
                # use a custom callback to handle the output from subprocesses
                outputCallback("\n".join(sorted(outputBuffer)))
        #print("READY!")
        resultDirs = resultDirs.get()

    summary = f"CONTIN analysis with {nthreads} thread{'s' if nthreads > 1 else ''} took {time.time()-start:.1f}s."
    return [rd for rd in resultDirs if rd is not None], summary

def getValueDictFromLines(lines, **kwargs):
    """Searches the given list of lines for the keys of the given arguments
    and converts the values to float. Returns the completed dict."""
    # search begin of common variables
    lstart = [idx for idx, line in enumerate(lines)
                  if "INPUT DATA FOR CHANGES TO COMMON VARIABLES" in line]
    # search end of common variables section (where the next begins)
    lend   = [idx for idx, line in enumerate(lines)
                  if "FINAL VALUES OF CONTROL VARIABLES" in line]
    result = dict()
    if len(lstart) and len(lend):
        result = {key: float(line.split()[-1])
                    for line in lines[lstart[0]:lend[0]]
                    for key, pattern in kwargs.items() if pattern in line}
    return result

def getContinUserVars(lines):
    """Extract previously set user variables for environmental values
    from CONTIN output data.
    *lines*: List of lines of CONTIN output data."""
    varmap = getValueDictFromLines(lines,
                temp="RUSER    18", angle="RUSER    17", visc="RUSER    19",
                refrac="RUSER    15", wavelen="RUSER    16", score="RUSER    11")
    # convert to SI units
    varmap["visc"] *= 1e-3
    varmap["wavelen"] *= 1e-9
    varmap["gamma"] = getDLSgammaSi(varmap["angle"], varmap["refrac"], varmap["wavelen"],
                                    varmap["temp"], varmap["visc"])
    return varmap

def getLineNumber(lines, phrases, debug=False):
    """Returns the line numbers containing the provided phrases after searching
    for the previous phrases sequentially. Search starts with the first phrase,
    once it is found, search starts with the 2nd phrase from that line,
    until the last phrase is found. Ignores early matches of the final phrase."""
    nums = []
    for i, line in enumerate(lines):
        if phrases[len(nums)] in line:
            if debug:
                print("found '{}' on line {}.".format(phrases[len(nums)], i))
            nums.append(i)
            if len(phrases) == len(nums):
                return nums
    return nums

def getContinInputCurve(inputAbsPath):
    assert inputAbsPath.is_file()
    # read in line by line, some adjustments required for parsing floats
    startLine, count = 0, 0
    with open(inputAbsPath) as fd:
        startLine, count = [(idx, int(line.split()[-1]))
                            for idx, line in enumerate(fd) if "NY" in line][0]
    lines = []
    with open(inputAbsPath) as fd:
        lines = fd.readlines()
    return [float(f) for line in lines[startLine+1:] for f in line.split()][count:]

def getContinResults(sampleDir, angle=None):
    """*sampleDir*: A pathlib Path of the location where the CONTIN results can be found."""
    sampleDir = Path(sampleDir)
    # check first if there was any CONTIN output generated
    resultsFile = sampleDir / OutputFn
    if not resultsFile.is_file():
        # try the subdir first
        assert angle is not None, "An angle in degrees has to be provided"
        resultsFile = sampleDir / getContinOutputDirname(float(angle)) / OutputFn
        if not resultsFile.is_file():
            print("No distribution found in\n '{}'!".format(resultsFile.parent))
            return None, None, None
    # read in line by line, some adjustments required for parsing floats
    lines = []
    with open(resultsFile) as fd:
        lines = fd.readlines()
    # find the beginning and end of the fitted correlation curve
    startLines = getLineNumber(lines, ["T            Y", "0PRECIS"])
    if not len(startLines):
        print(f"Fitted curve not found in CONTIN output!\n ({resultsFile})")
        return None, None, None
    dfStart, dfEnd = startLines[-2]+1, startLines[-1]
    dfFit = pd.DataFrame([f for line in lines[dfStart:dfEnd] for f in grouper(line.split(), 2)],
                         columns=('tau', 'corrFit'), dtype=float)
    dfFit.corrFit = dfFit.corrFit**2 # to be compared with measured data
    # get input correlation curve first, to be added to fitted correlation curve
    dfFit['corrIn'] = getContinInputCurve(sampleDir/InputFn)
    # find the beginning and end of the distribution data
    startLines = getLineNumber(lines, ["CHOSEN SOLUTION", "ORDINATE"])
    if not len(startLines):
        print(f"Distribution data not found in CONTIN output!\n ({resultsFile})")
        return None, None, None
    gridSize = int(getValueDictFromLines(lines, distribSize="NG        0").get('distribSize',0))
    dfStart = startLines[1]+1
    lineEnd = 31 # do not parse floats beyond this column
    # convert CONTIN output distrib to parseable data for pandas
    fixedFloatFmt = io.StringIO("\n".join([line[:lineEnd].replace("D", "E")
                                for line in lines[dfStart:dfStart+gridSize]]))
    dfDistrib = pd.read_csv(fixedFloatFmt, delim_whitespace=True,
                            names=("distrib", "err", "decay"))
    dfDistrib = dfDistrib[["decay", "distrib", "err"]] # reorder to (x,y,u)
    # update x/abscissa with values from another section of the output
    # to avoid duplicates due to low precision in solution output parsed above
    startLines = getLineNumber(lines, ["GRID POINT"])
    if len(startLines):
        dfStart = startLines[0]+1
        decayNew = np.fromiter([line.split()[0] for line in lines[dfStart:dfStart+gridSize]], float)
        dfDistrib.decay = decayNew
    varmap = getContinUserVars(lines)
    # parse original input data filename as well, if available
    # see genContinInput() for creating storedFn
    storedFn = lines[0][52:].strip()
    infn = sampleDir.parent / storedFn
    if storedFn[0] == '.': # starts with a dot
        infn = sampleDir.parent.parent / (sampleDir.parent.name + storedFn)
    varmap['dataFilename'] = infn
    return dfDistrib, dfFit, varmap


dlshelpers.py
# -*- coding: utf-8 -*-
# utils.py

import os, re, io, codecs
from warnings import warn
from pathlib import Path
from dateutil.parser import parse as parseDateTime
import zipfile
import bson
import pprint
from uuid import UUID

import numpy as np
import pandas as pd
from jupyter_analysis_tools.utils import isList

# transmission levels of ALV-CGS4F/8F given in measurement software, angles/detectors 1-4
transmissionLevels = np.array(((1., .1, .03, .01),)+((1., .3, .1, .03),)*3)

def angleToQ(deg, refrac, wavelen):
    return 4.*np.pi*refrac/wavelen*np.sin(deg*np.pi/360.)

# constant for the experiment at each angle
def calcDLSconst(q, temp, visc):
    kB = 1.38066E-23 # Boltzman constant, [kB] = Newton m / Kelvin
    return kB*temp/(6*np.pi*visc)*q**2

def getDLSgammaSi(angle, refrac, wavelen, temp, visc):
    return calcDLSconst(angleToQ(angle, refrac, wavelen), temp, visc)

def parseValue(val):
    val = val.strip('"')
    try:
        val = float(val)
    except ValueError:
        pass
    return val

def bufferFromLines(ln):
    assert isList(ln)
    return io.StringIO("".join(ln))

def readDLSMetaASC(lines, encoding='cp1250'):
    """Reads the measurement settings (temperature, viscosity, refractive index and wavelength).
    The *MeanCR* field is ignored because they were found to be either inconsistent with the
    recorded count rate in the same file (type `ALV-7004/USB`) or they have the wrong ordering
    compared to the recorded count rate in the same file (type `ALV-7004 CGS-8F Data`). The
    means can be calculated easily from the count rate read by *readDLSDataASC()*.
    **For examples**, please see the tests for this function in `tests/test_dlshelpers.py`.
    """
    assert isList(lines)
    meta = {'filetype': lines[0].strip()}
    # section names to look for and their keys in the metadata dict
    sections = [('corrStartLn', '"Correlation"'), ('crStartLn', '"Count Rate"'),
                ('monitorDiodeLn', 'Monitor Diode'), ('cum1StartLn', '"Cumulant 1.Order"'),
                ('cum2StartLn', '"Cumulant 2.Order"'), ('cum3StartLn', '"Cumulant 3.Order"'),
                ('stddevStartLn', '"StandardDeviation"')]
    meta['sections'] = {key: -1 for key, _ in sections}
    key, pattern = sections.pop(0)
    for idx, line in enumerate(lines):
        if pattern in line:
            meta['sections'][key] = idx
            if len(sections):
                key, pattern = sections.pop(0)
            else:
                break
    assert meta['sections']['corrStartLn'] > 0, \
            'The mandatory "Correlation" section was not found!'
    # read the recorded environment data, device settings and sample notes
    df = pd.read_csv(bufferFromLines(lines), sep=':', encoding=encoding,
                       skiprows=1, nrows=meta['sections']['corrStartLn']-2,
                       index_col=0,
                       names=range(5), # also read lines containing more colons = multiple fields
                       na_filter=False, # don't replace empty fields with NaN
                       header=None,
                       #quotechar='"',
                       escapechar="\t", # filters any TAB whitespace
                       skipinitialspace=True # strips whitespace from fields outside of quotes
                      )
    # parse floats and put datetimes back together
    meta.update({name.strip():
                 parseValue(':'.join(field for field in df.T[name] if len(field)).strip())
                 for name in df.index
                 if "MeanCR" not in name
                })
    if meta['sections']['monitorDiodeLn'] > 0:
        meta['monitorDiode'] = float(lines[meta['sections']['monitorDiodeLn']].split()[-1])
    # return the sorted dict
    return meta

def readDLSDataASC(filename, showProgress=False, encoding='cp1250',
                   attenKeys={'key': 'abgeschwächt', 'detectorKey': 'detektor', 'levelKey': 'stufe'}):
    """
    *Mode*: The mode describes which channels (count rate columns) were used for auto- or
    cross-correlation, as given in
    [the specs sheet by ALV](https://www.alvgmbh.de/dwnload/alv7000specs.pdf).
    *corr*: Data with `C-CH0/1+1/0` mode contain the cross-correlation of two channels
            in both directions in the first and second column. Both columns are averaged
            on load and the averaged column is prepended and indexed with the measurement
            angle, to be used for further processing (improves data quality, see issue #1).
    **For examples**, please see the tests for this function in `tests/test_dlshelpers.py`.
    """
    if showProgress:
        print('.', end="") # some progress output
    filename = Path(filename)
    filelines = []
    with codecs.open(filename, encoding=encoding) as fh:
        filelines = [ln for ln in fh.readlines() if len(ln.strip())] # filter empty lines
    data = readDLSMetaASC(filelines)
    data['filename'] = filename
    data['timestamp'] = parseDateTime(data['Date']+' '+data['Time'])

    # parsing the scattering angles
    angles = [value for key,value in data.items() if key.startswith("Angle")]
    data['angles'] = angles

    # try to get the concentration from the memo field, TODO: write a test for this
    data['memo'] = "".join([value for key,value in data.items() if key.startswith("SampMemo")])
    #print(f"memostr: '{data['memo']}'")
    data['memofields'] = [field.strip(' ,;') for field in data['memo'].split()]
    try:
        concentration = [float(field.split(':')[1])
                         for field in data['memofields'] if ':' in field][0]
        concentration = 1/float(concentration)
    except (IndexError, ValueError):
        concentration = 1
    data['concentration'] = concentration

    # parse attenuation values
    atten = [value for key,value in data.items() if key.startswith("Atten")]
    atten += [1.0] * (len(angles) - len(atten)) # pad attenuation list to have same amount as angles (only 3 stored in file)
    if attenKeys['key'] in data['memo']:
        attenManual = getAttenuationFromMemo(data['memo'], len(angles), **attenKeys)
        if atten != attenManual:
            warn("\n"f"  Attenuation values specified in sample description: {attenManual}\n"
                 f"  ('{data['memo']}'),\n"
                 f"  differ from the values stored by software:          {atten}!\n"
                 f"  -> Using values from sample description:            {attenManual}.", # 'cause they may contain 4 instead of 3 values
                 None, stacklevel=2)
            atten = attenManual
    data['attenuation'] = pd.DataFrame(np.array(atten).reshape(1,-1), columns=angles)

    data = dict(sorted(data.items(), key=lambda x: x[0].lower()))

    # read a section of data columns from header until begin of next section, or file end
    lineIndices = np.array([idx for key, idx in data['sections'].items()], dtype=int)
    def sectionEndLn(startLn): # return the index of any section directly following the given index
        followingSections = lineIndices[lineIndices > startLn]
        return followingSections.min() if followingSections.size else len(filelines)

    # read correlation data columns (mandatory)
    startln, endln = data['sections']['corrStartLn']+1, sectionEndLn(data['sections']['corrStartLn'])
    corr = pd.read_csv(bufferFromLines(filelines), encoding=encoding, sep=r'\s+',
                       index_col=0, header=None,
                       skiprows=startln, nrows=endln-startln)
    if data['Mode'].strip() == "C-CH0/1+1/0": # cross-correlation of two channels in both directions
        corr.insert(0, 0, corr.iloc[:,:2].mean(axis=1))

    # adjust column names if more columns than angles were found
    columns = angles
    if len(corr.columns) > len(columns):
        # repeat the last angle with suffix if there are more columns
        columns = columns + ([columns[-1],] * (len(corr.columns) - len(columns)))
        columns = [f"{int(angle)}_{idx}" if idx else angle for idx, angle in enumerate(columns)]
    corr.rename(columns=dict(zip(corr.columns, columns)), inplace = True)
    corr.index.names = ["tau"]
    data['correlation'] = corr
    
    # read the count rate data columns
    if data['sections']['crStartLn'] > 0: # does not exist in an 'averaged' data file
        startln, endln = data['sections']['crStartLn']+1, sectionEndLn(data['sections']['crStartLn'])
        cr = pd.read_csv(bufferFromLines(filelines), encoding=encoding, sep=r'\s+',
                         #names=["time"]+columns,
                         header=None,
                         index_col=0,
                         skiprows=startln, nrows=endln-startln)
        if data['Mode'].strip() == "C-CH0/1+1/0": # cross-correlation of two channels in both directions
            cr.insert(0, 0, cr.iloc[:,:2].mean(axis=1))
        cr.rename(columns=dict(zip(cr.columns, columns)), inplace = True)
        cr.index.names = ["time"]
        data['countrate'] = cr

    # read cumulants from data
    cumulants = []
    for num in range(3):
        cumulants.append(None)
        secname = f"cum{num+1}StartLn"
        if data['sections'][secname] <= 0:
            continue
        startln, endln = data['sections'][secname]+1, sectionEndLn(data['sections'][secname])
        secname = filelines[startln-1].strip().strip('"')
        cumulants[num] = pd.read_csv(bufferFromLines(filelines), encoding=encoding, sep=r'\t',
                                     names=("value",), index_col=0, engine='python',
                                     skiprows=startln, nrows=endln-startln)
    if any([cumu is not None for cumu in cumulants]):
        data['cumulants'] = cumulants

    # read the standard deviation
    stddevStartLn = data['sections']['stddevStartLn']
    if stddevStartLn > 0:
        startln, endln = stddevStartLn+1, sectionEndLn(stddevStartLn)
        data['stddev'] = pd.read_csv(bufferFromLines(filelines), encoding=encoding, sep=r'\s+',
                                     names=["tau", columns[0]], index_col=0,
                                     skiprows=startln, nrows=endln-startln)
    return data

def getAttenuationFromMemo(memo, count=4, detectorKey='detektor', levelKey='stufe', key=None):
    """Extract manually specified attenuation of individual scattering angles from a given
    measurement data files (ASC) *SampMemo* field.
    Different notations and numbers of detectors are accepted (test cases):

    >>> getAttenuationFromMemo("Toluolmessung für Statik,, Detektor 1 und 2, 1 Stufe abgeschwächt")
    [0.1, 0.3, 1.0, 1.0]

    >>> getAttenuationFromMemo("Toluolmessung für Statik,, Detektor 1 und 2, 1 Stufe abgeschwächt, Detektor 3 , 2 Stufen abgeschwächt")
    [0.1, 0.3, 0.1, 1.0]

    >>> getAttenuationFromMemo("Toluolmessung für Statik,, Detektor 1 und 2, 1 Stufe abgeschwächt, Detektor 3 und 4, 2 Stufen abgeschwächt")
    [0.1, 0.3, 0.1, 0.1]

    >>> getAttenuationFromMemo("Toluolmessung für Statik,, Detektor 1, 1 Stufe abgeschwächt, Detektor 3 , 2 Stufen abgeschwächt")
    [0.1, 1.0, 0.1, 1.0]

    >>> getAttenuationFromMemo("1:100 verdünnt mit Wasser, ungefiltert,, Detektor 1(10%) und 2(30%), 1 Stufe abgeschwächt")
    [0.1, 0.3, 1.0, 1.0]

    Table of transmission levels 1-4 (columns) as defined in the program for detectors 1-4 (rows):

    >>> print(transmissionLevels)
    [[1.   0.1  0.03 0.01]
     [1.   0.3  0.1  0.03]
     [1.   0.3  0.1  0.03]
     [1.   0.3  0.1  0.03]]
    """
    def getnumbers(text, idxoffset=0):
        return ([(int(idx)+idxoffset, int(inner)/100 if len(inner) else 1.)
                for idx, _, inner in re.findall(r"\b(\d)\b(\((\d+)%\))?", text)])
    atten, detector = list(np.ones(count)), ()
    for field in memo.split(','):
        field = field.strip().lower()
        #print(f"-> field '{field}':")
        # extract detectors indices first, may contain transmission % already
        if detectorKey in field:
            detector = getnumbers(field, idxoffset=-1)
            #print(detector)
        # extract transmission level next, level indices map to hard coded transmission factor table
        if levelKey in field:
            level = getnumbers(field)
            detector = [(d, transmissionLevels[d,level[d%len(level)][0]]) for d, _ in detector]
            #print(detector)
        for d, val in detector:
            atten[d] = val
    return atten

def convertAPKWentries(data):
    """Gets a data dictionary loaded from an APKW file and adds or converts field
    values for compatibility with existing code."""
    data['timestamp'] = data['MetaInfo']['StartTime']
    data['Samplename'] = data['InputParameter']['Name']
    data['memo'] = data['InputParameter']['Comment']
    data['memofields'] = [field.strip(' ,;') for field in data['memo'].split()] # same a in getDLSFileData()
    data['concentration'] = 1
    # use the ref. index value of the solvent
    data['Refractive Index'] = data['InputParameter']['Solvent']['RefractiveIndex']
    if data['Refractive Index'] == 0.:
        # find the ref. index in look-up table if not set
        data['Refractive Index'] = [entry['Value']
             for entry in data['InputParameter']['Solvent']['RefractiveIndices']
             if     entry['Key']['Temperature'] == data['InputParameter']['Solvent']['Temperature']
                and entry['Key']['Wavelength'] == data['InputParameter']['Solvent']['Wavelength']][0]
    data['Wavelength [nm]'] = data['InputParameter']['Solvent']['Wavelength']*1e9
    data['Temperature [K]'] = data['InputParameter']['Solvent']['Temperature']
    data['Viscosity [cp]'] = data['InputParameter']['Solvent']['Viscosity']*1e3
    if data['MetaInfo']['InstrumentType'] == 'Litesizer 100':
        data['Angle [°]'] = 175.
        data['angles'] = [data['Angle [°]']]
    data['correlation'] = pd.DataFrame(data['RawData'][0]['CorrelationDataScaled'],
                                       columns=data['angles'],
                                       index=data['RawData'][0]['CorrelationDelayTimes'],)
    data['correlation'] -= 1.
    data['correlation'].index *= 1e3
    data['correlation'].index.names = ["tau"]
    data['countrate'] = pd.DataFrame(data['RawData'][0]['IntensityTrace'], columns=data['angles'])
    data['countrate'] *= 10. # converted to match the exported XLS sheet data
    del data['RawData'] # remove obsolete, duplicate data
    data['Duration [s]'] = data['InputParameter']['MeasurementTime']
    # hardcoded count rate reading times? nowhere stored in file
    data['countrate'].index = np.linspace(0.1, 0.1*data['countrate'][data['Angle [°]']].size,
                                          data['countrate'][data['Angle [°]']].size)
    data['attenuation'] = pd.DataFrame({data['Angle [°]']: (data['Attenuation'],)})
    # extract data of the cumulant fit
    rr = data['RecentRun']
    data['cumulantfit'] = pd.DataFrame(zip(rr['CorrelationFitFunction'], rr['G2Residuals']),
                                       columns=('CorrelationFitFunction','G2Residuals'),
                                       index=rr['CumulantDelayTimes'])
    data['cumulantfit']['CorrelationFitFunction'] -= 1.
    data['cumulantfit'].index.names = ["tau"]
    # remove obsolete, duplicate data
    del data['RecentRun']['CorrelationFitFunction']
    del data['RecentRun']['G2Residuals']
    del data['RecentRun']['CumulantDelayTimes']
    # extract reconstructed distribution
    data['distribution'] = pd.DataFrame(zip(rr['ReconstructionIntensityDistribution'], rr['ReconstructionVolumeDistribution'], rr['ReconstructionNumberDistribution']),
                                       columns=('intensity', 'volume', 'number'),
                                       index=rr['ReconstructionRadiusNodes'])
    data['distribution'].index.names = ["radius"]
    # remove obsolete, duplicate data
    del data['RecentRun']['ReconstructionIntensityDistribution']
    del data['RecentRun']['ReconstructionVolumeDistribution']
    del data['RecentRun']['ReconstructionNumberDistribution']
    del data['RecentRun']['ReconstructionRadiusNodes']
    return data

def readDLSDataAPKW(filename):
    """Yields none or more unique dicts, one for each measurement,
    stored in the given APKW file name."""
    #print(filename)
    filename = Path(filename)
    if zipfile.is_zipfile(filename):
        with zipfile.ZipFile(filename, 'r') as zf:
            for zi in zf.infolist():
                if zi.filename.startswith('measurement'):
                    #print(zi)
                    data = bson.loads(zf.read(zi))
                    #pprint.pprint(data)
                    if data['StorageStatus'] == 3 and data['UsedAngle'] > 0:
                        data['filename'] = filename / str(data['Id']) # remember it here
                        yield convertAPKWentries(data)

def readDLSData(files, *args, **kwargs):
    """Read DLS measurements from provided files, *.ASC or *.apkw are supported."""
    if not isinstance(files, list) and not isinstance(files, tuple):
        files = (files,)
    # gather all measurement data from APKW files first
    dirData = list({datadict['Id']: datadict
                    for fn in files
                    for datadict in readDLSDataAPKW(fn)}.values())
    # try to read ASC files
    if not len(dirData):
        dirData = [readDLSDataASC(fn) for fn in files]
    return dirData

if __name__ == "__main__":
    import doctest
    doctest.testmod()
